Add PostgreSQL-based metastore.

mosuka opened the pull request:
### Context / purpose
Add PostgreSQL-based metastore.
Close #389

### Description
Add PostgreSQL-based metastore.

### How was this PR tested?
```
$ cd quickwit-metastore

$ docker-compose up -d --force-recreate

$ cargo test --features=postgresql

$ docker-compose down
```

### Checklist
*Remove or complete this list if required.*
- [x] tested locally
- [x] added unit tests

mosuka commented:
@fulmicoton @fmassot 
I have a question on how to specify the max connection pool size.

Currently, the default value is specified, but I think it would be better to allow users to freely set the pool size.
However, in the current CLI, there is no way to pass the metastore-specific configuration values other than specifying the connection destination with `--metastore-uri`.

So, came up with the following methods:

Add an environment variable
```
POSTGRES_MAX_CONNECTION_POOL_SIZE=50
```

Extend the connection URI specified by `--metastore-uri`
```
postgres://quickwit-dev:quickwit-dev@localhost/quickwit-dev?max_connection_pool_size=50
```

Do you have any ideas to pass the metastore specific configurations?

fulmicoton commented on quickwit-metastore/migrations/2021-08-29-122131_create_splits/up.sql:
```diff
@@ -0,0 +1,16 @@
+CREATE TABLE splits (
+    split_id VARCHAR(50) PRIMARY KEY, 
+    split_state INTEGER NOT NULL, 
```
Let's use strings instead of INTEGER for the split state.

PostgreSQL also has enum that supposedly offer the best of two worlds, but they are a pain to modify.

mosuka replied:
I changed to use VARCHAR instead of INTEGER for split_state.

fulmicoton commented on quickwit-metastore/migrations/2021-08-29-122131_create_splits/up.sql:
```diff
@@ -0,0 +1,16 @@
+CREATE TABLE splits (
+    split_id VARCHAR(50) PRIMARY KEY, 
+    split_state INTEGER NOT NULL, 
+    num_records BIGINT NOT NULL, 
+    size_in_bytes BIGINT NOT NULL, 
+    start_time_range BIGINT,
+    end_time_range BIGINT,
+    generation BIGINT NOT NULL,
+    update_timestamp BIGINT NOT NULL, 
```
Having a strict schema is can be great, but we will have more than one implementation, but as I commented on notion, we should rely on JSON here for most of this data.

We want to have as a field:

split_id,
index_id,
split_state,
start_time_range
end_time_range

Possibly tags as well, (which would then be an array).
The rest should be in a JSON field.

mosuka replied:
I left the following columns in the table and added a new `json` column.
The `json` column now contains a JSON string of `SplitMetadataAndFooterOffsets`.
This is so that the string in the `json` column can be easily deserialized to `SplitMetadataAndFooterOffsets`.

```
CREATE TABLE splits (
    split_id VARCHAR(50) PRIMARY KEY,
    split_state VARCHAR(30) NOT NULL,
    start_time_range BIGINT,
    end_time_range BIGINT,
    tags TEXT NOT NULL,
    json TEXT NOT NULL,
    index_id VARCHAR(50) NOT NULL,

    FOREIGN KEY(index_id) REFERENCES indexes(index_id)
);
```

fulmicoton commented on quickwit-metastore/src/error.rs:
```diff
@@ -30,6 +30,9 @@ use crate::checkpoint::IncompatibleCheckpoint;
 #[derive(Debug, Error)]
 pub enum MetastoreError {
+    #[error("Connection error: `{message}`.")]
+    ConnectionError { message: String },
```
good idea

fulmicoton commented on quickwit-metastore/src/metastore.rs:
```diff
@@ -130,6 +132,28 @@ impl Default for SplitState {
 }
 
+impl FromPrimitive for SplitState {
+    fn from_i64(num_split_state: i64) -> Option<SplitState> {
+        match num_split_state {
+            0 => Some(SplitState::New),
+            1 => Some(SplitState::Staged),
+            2 => Some(SplitState::Published),
+            3 => Some(SplitState::ScheduledForDeletion),
+            _ => None,
+        }
```
This is useless... 
- From<u64> should be sufficient
- if From<i64> was a good thing, it could still call from_u64.
- anyway let's use strings for enum.

fulmicoton commented on quickwit-metastore/src/model.rs:
```diff
@@ -0,0 +1,170 @@
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+use std::ops::{Range, RangeInclusive};
```
we should put that in a postgresql module.  (as well as all of the files specific to postgres in this PR)

`quickwit-metastore/src/postgres/...`

Right now it is confusing...

mosuka replied:
I did not fully understand the use of diesel.rs. I've fixed it now.

fulmicoton commented:
@mosuka For the moment, I think you can keep that a hardcoded constant. We have a lot of such constants and I think we need to think deeply about how to let the user configure them. (probably a config file)

mosuka commented:
@fulmicoton 
I've reflected on your feedback. Please take another look if you have time.
Thanks.

fulmicoton commented on quickwit-metastore/src/metastore.rs:
```diff
@@ -20,15 +20,18 @@
 
 use async_trait::async_trait;
 use chrono::Utc;
+// use num::FromPrimitive;
```
remove

fulmicoton commented on quickwit-metastore/src/metastore/postgresql_metastore.rs:
```diff
@@ -0,0 +1,1139 @@
+fn establish_connection(
+    database_uri: &str,
+) -> anyhow::Result<Pool<ConnectionManager<PgConnection>>> {
+    dotenv().ok();
```
let's remove the use to dotenv.

fulmicoton commented on quickwit-metastore/src/metastore/postgresql_metastore.rs:
```diff
@@ -0,0 +1,1139 @@
+
+    let mut retry_cnt = 0;
+
+    loop {
```
make it a for loop

fulmicoton commented on quickwit-metastore/src/metastore/postgresql_metastore.rs:
```diff
@@ -0,0 +1,1139 @@
+                }
+
+                retry_cnt += 1;
+                continue;
```
continue is useless here.

fulmicoton commented on quickwit-metastore/src/metastore/postgresql_metastore.rs:
```diff
@@ -0,0 +1,1139 @@
+/// Get the PostgreSQL-based metastore. The metastore is created as a singleton.
+pub async fn get_postgresql_metastore(database_uri: &str) -> &'static PostgresqlMetastore {
+    static POSTGRESQL_METASTORE: OnceCell<PostgresqlMetastore> = OnceCell::const_new();
+    POSTGRESQL_METASTORE
```
This is incorrect.

It cannot be a single if it has an argument.

fulmicoton replied:
HashMap<String, Weak<PostgresqlMetastore>> would be better.

fulmicoton commented on quickwit-metastore/src/metastore/postgresql_metastore.rs:
```diff
@@ -0,0 +1,1139 @@
+const MAX_CONNECTION_POOL_SIZE: u32 = 10;
+const CONNECTION_TIMEOUT: u64 = 10;
+const MAX_CONNECTION_RETRY_COUNT: u32 = 10;
+const CONNECTION_STATUS_CHECK_INTERVAL: u64 = 2;
```
You can have const duration.

fulmicoton commented on quickwit-metastore/src/metastore/postgresql_metastore.rs:
```diff
@@ -0,0 +1,1139 @@
+embed_migrations!("migrations/postgresql");
+
+const MAX_CONNECTION_POOL_SIZE: u32 = 10;
+const CONNECTION_TIMEOUT: u64 = 10;
```
you can have const Duration

fulmicoton commented on quickwit-metastore/migrations/postgresql/2021-09-02-060554_create_splits/up.sql:
```diff
@@ -0,0 +1,11 @@
+CREATE TABLE splits (
+    split_id VARCHAR(50) PRIMARY KEY,
+    split_state VARCHAR(30) NOT NULL,
+    start_time_range BIGINT,
```
I think it is ok not to have an index on start_time_ragne, end_time_range, and tags at the moment.

fulmicoton commented on quickwit-metastore/migrations/postgresql/2021-09-02-060554_create_splits/up.sql:
```diff
@@ -0,0 +1,11 @@
+    split_state VARCHAR(30) NOT NULL,
+    start_time_range BIGINT,
+    end_time_range BIGINT,
+    tags TEXT NOT NULL,
```
tags could be an array field.
https://www.postgresql.org/docs/9.1/arrays.html

fulmicoton commented on quickwit-metastore/migrations/postgresql/2021-09-02-060554_create_splits/up.sql:
```diff
@@ -0,0 +1,11 @@
+    start_time_range BIGINT,
+    end_time_range BIGINT,
+    tags TEXT NOT NULL,
+    json TEXT NOT NULL,
```
rename `split_metadata_json`  and comment that it contains the entire split metadata.

fulmicoton commented on quickwit-metastore/src/metastore/postgresql_metastore.rs:
```diff
@@ -0,0 +1,1139 @@
+#[async_trait]
+impl crate::tests::test_suite::DefaultForTest for PostgresqlMetastore {
+    async fn default_for_test() -> Self {
+        use std::env;
```
Can you put instruction somewhere about how to run tests locally?
Typically a docker compose file spinning a postgres and environment variable would be great.

dotenv with a committed .env file pointing at the docker-compose instance is ok, but please make the dotenv a dev-dependency.

fulmicoton commented on quickwit-metastore/src/postgresql.rs:
```diff
@@ -0,0 +1,4 @@
+#[allow(missing_docs)]
+pub mod model;
```
postgresql/mod.rs please

fulmicoton commented on quickwit-metastore/src/lib.rs:
```diff
@@ -31,13 +31,22 @@ etc.
 mod metastore;
 mod metastore_resolver;
+#[allow(missing_docs)]
+pub mod postgresql;
```
#[feature(postresql)]

fulmicoton commented on quickwit-metastore/src/metastore_resolver.rs:
```diff
@@ -68,6 +69,7 @@ impl Default for MetastoreUriResolver {
             .register("file", SingleFileMetastoreFactory::default())
             .register("s3", SingleFileMetastoreFactory::default())
             .register("s3+localstack", SingleFileMetastoreFactory::default())
+            .register("postgresql", PostgresqlMetastoreFactory::default())
```
```
if cfg!(feature="postgresql") {
   // ...
}
```

fulmicoton commented on quickwit-metastore/src/metastore/postgresql_metastore.rs:
```diff
@@ -0,0 +1,1120 @@
+
+        match initialize_db(&*connection_pool) {
+            Ok(_) => info!("Database initialized"),
+            Err(err) => error!("Failed to initialize database {:?}", err),
```
We should return an error here.

mosuka replied:
Sorry, I'll fix it.

mosuka replied:
Fixed.

fulmicoton commented on quickwit-metastore/src/metastore/postgresql_metastore.rs:
```diff
@@ -0,0 +1,1120 @@
+}
+
+impl PostgresqlMetastore {
+    /// Creates a meta store given a storage.
```
the comment is wrong.

mosuka replied:
Sorry, I'll fix it.

mosuka replied:
Fixed.

fulmicoton commented on quickwit-metastore/src/metastore/postgresql_metastore.rs:
```diff
@@ -0,0 +1,1120 @@
+            }
+        })?);
+
+        let connection_pool_clone = connection_pool.clone();
```
is cloning required here?

mosuka replied:
This cloning was not necessary...

mosuka replied:
Fixed.

fulmicoton commented on quickwit-metastore/src/metastore/postgresql_metastore.rs:
```diff
@@ -0,0 +1,1120 @@
+            if conn_res.is_ok() {
+                break;
+            }
+            tokio::time::sleep(CONNECTION_STATUS_CHECK_INTERVAL).await;
```
Can we return an error if the connection could not be established after 3 attempts?

mosuka replied:
Fixed.

fulmicoton commented on quickwit-metastore/src/metastore/postgresql_metastore.rs:
```diff
@@ -0,0 +1,1120 @@
+                })?;
+
+        // Apply checkpoint_delta
+        index_metadata
```
I'd rather have the checkpoint check logic to be part of the transaction...

mosuka replied:
If I move the checkpoint logic to a transaction, I would expect to roll back the transaction if an error occurs in `Checkpoint::try_apply_delta`.
In `transaction` functions, it seems that if I want to roll back with user code other than DB operations, I have to return `diesel::result::Error:.RollbackTransaction`.
Then the `MetastoreError::IncompatibleCheckpointDelta` that the `publish_splits` function has to return cannot be known outside the transaction.
Is there a good way to know non-diesel::result::Error that occurs in a transaction outside of the transaction?

fulmicoton commented on quickwit-metastore/src/metastore/postgresql_metastore.rs:
```diff
@@ -0,0 +1,1120 @@
+
+        // Check for the inclusion of non-publishable split IDs.
+        // Except for SplitState::Staged and SplitState::Published, you cannot publish.
+        let non_publishable_split_ids = self
```
That logic is not bad but is redundant with the logic existing within the transaction. I think we can remove it and make our error more specific in the transaction code.

fulmicoton commented on quickwit-metastore/src/metastore/postgresql_metastore.rs:
```diff
@@ -0,0 +1,1120 @@
+                        break;
+                    }
+                }
+                return Err(diesel::result::Error::NotFound);
```
We are finding error_split_id and then not using it and returning an error.

fulmicoton commented on quickwit-metastore/src/metastore/postgresql_metastore.rs:
```diff
@@ -0,0 +1,1120 @@
+
+            // Find out the ID of the update error and return an error if there is one.
+            if mark_as_deleted_split_ids.len() < replaced_split_ids.len() {
+                for split_id in replaced_split_ids.iter() {
```
Editing split_error and returning an error, that will be converted afterwards is a very ugly pattern. If you have ideas for refactoring, it might be better.

fulmicoton commented on quickwit-metastore/src/metastore/postgresql_metastore.rs:
```diff
@@ -0,0 +1,1120 @@
+
+            // The tags are empty (unspecified) or the tags are specified
+            // and the split has a matching tag, it will be added to the list.
+            let mut match_tag = false;
```
Can you that logic in an external function?

/// ...
fn does_match_tag_filter(tags: &[String], tag_filter: &[String]) -> bool {
}


it could be factorized in the single file metastore too.

mosuka replied:
https://github.com/quickwit-inc/quickwit/blob/8b460dfd684912e3668f8192e5a9d96edd4fbcaa/quickwit-metastore/src/metastore/postgresql_metastore.rs#L811-L819

https://github.com/quickwit-inc/quickwit/blob/8b460dfd684912e3668f8192e5a9d96edd4fbcaa/quickwit-metastore/src/metastore/single_file_metastore.rs#L386-L393

fulmicoton commented:
The following tests are failing for me.
```

---- metastore::postgresql_metastore::common_tests::test_metastore_delete_splits stdout ----
ng
thread 'metastore::postgresql_metastore::common_tests::test_metastore_delete_splits' panicked at 'assertion failed: matches!(result, MetastoreError :: IndexDoesNotExist { .. })', quickwit-metastore/src/tests.rs:897:13

---- metastore::postgresql_metastore::common_tests::test_metastore_split_update_timestamp stdout ----
ng
thread 'metastore::postgresql_metastore::common_tests::test_metastore_split_update_timestamp' panicked at 'called `Result::unwrap()` on an `Err` value: InternalError { message: "Failed to create index", cause: relation "indexes" does not exist }', quickwit-metastore/src/tests.rs:1685:14

---- metastore::postgresql_metastore::common_tests::test_metastore_replace_splits stdout ----
ng
thread 'metastore::postgresql_metastore::common_tests::test_metastore_replace_splits' panicked at 'assertion failed: matches!(result, MetastoreError :: IndexDoesNotExist { .. })', quickwit-metastore/src/tests.rs:631:13

---- metastore::postgresql_metastore::common_tests::test_metastore_stage_split stdout ----
ng
thread 'metastore::postgresql_metastore::common_tests::test_metastore_stage_split' panicked at 'assertion failed: matches!(result, MetastoreError :: IndexDoesNotExist { .. })', quickwit-metastore/src/tests.rs:188:9


failures:
    metastore::postgresql_metastore::common_tests::test_metastore_delete_splits
    metastore::postgresql_metastore::common_tests::test_metastore_replace_splits
    metastore::postgresql_metastore::common_tests::test_metastore_split_update_timestamp
    metastore::postgresql_metastore::common_tests::test_metastore_stage_split

```

mosuka commented:
@fulmicoton Let me check...

fulmicoton commented:
The test are passing correctly when run single threaded.
Please complete the test instructions, or tweak the DefaultForTest impl of PostgresMetastore to prevent concurrent execution of these tests.

mosuka commented:
@fulmicoton I tweaked the DefaultForTest implementation of PostgresMetastore to prevent concurrent execution of these tests.
https://github.com/quickwit-inc/quickwit/blob/8b460dfd684912e3668f8192e5a9d96edd4fbcaa/quickwit-metastore/src/metastore/postgresql_metastore.rs#L1104-L1134

mosuka commented:
@fulmicoton 
I corrected the points that were pointed out. Please take another look.

fulmicoton reviewed (changes requested):
I requested a couple of changes.

fulmicoton commented on quickwit-metastore/src/metastore/postgresql_metastore.rs:
```diff
@@ -0,0 +1,1066 @@
+                             split_id={:?} {:?}",
+                            model_split.split_id, err
+                        );
+                        continue;
```
You did not address my comments.. How is this ok to just log ?

mosuka commented:
@fulmicoton 
Thanks for the review.
The part about not returning an error is my mistake. I have fixed it.
I also fixed the other comments you made, so please check them.

fulmicoton commented on quickwit-metastore/src/metastore/postgresql_metastore.rs:
```diff
@@ -0,0 +1,1095 @@
+                        .iter()
+                        .map(String::as_str)
+                        .collect::<Vec<_>>(),
+                )?
```
```suggestion
                )?;
```

fulmicoton commented on quickwit-metastore/src/metastore/postgresql_metastore.rs:
```diff
@@ -0,0 +1,1095 @@
+                        .iter()
+                        .map(String::as_str)
+                        .collect::<Vec<_>>(),
+                )?
```
```suggestion
                )?;
```

fulmicoton commented on quickwit-metastore/src/metastore/postgresql_metastore.rs:
```diff
@@ -0,0 +1,1095 @@
+                        .iter()
+                        .map(String::as_str)
+                        .collect::<Vec<_>>(),
+                )?
```
```suggestion
                )?;
```
